ethereum/sidecar/bridge_out_client.go

```diff
@@ -0,0 +1,181 @@
+package sidecar
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/codec/types"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	"cosmossdk.io/log"
+	sdkmath "cosmossdk.io/math"
+	bridgetypes "github.com/mezo-org/mezod/x/bridge/types"
+)
+
+// BridgeOutGrpcClient enables gRPC communication with mezod validator node needed
+// for the bridge-out process.
+type BridgeOutGrpcClient struct {
+	mutex          sync.Mutex
+	requestTimeout time.Duration
+	connection     *grpc.ClientConn
+}
+
+func NewBridgeOutGrpcClient(
+	logger log.Logger,
+	serverAddress string,
+	requestTimeout time.Duration,
+	registry types.InterfaceRegistry,
+) (*BridgeOutGrpcClient, error) {
+	connection, err := grpc.NewClient(
+		serverAddress,
+		// TODO: Consider using TLS protocol so that the Mezo node and Ethereum
+		//       sidecar can be run on separate servers.
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithDefaultCallOptions(
+			grpc.ForceCodec(codec.NewProtoCodec(registry).GRPCCodec()),
+		),
+	)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"failed to create gRPC bridge-out client for Ethereum sidecar [%w]",
+			err,
+		)
+	}
+
+	c := &BridgeOutGrpcClient{
+		requestTimeout: requestTimeout,
+		connection:     connection,
+	}
+
+	go func() {
+		// Test the connection to the Ethereum sidecar bridge-out server by
+		// verifying we can successfully execute `GetAssetsUnlockedEntries`.
+		ctxWithTimeout, cancel := context.WithTimeout(
+			context.Background(),
+			requestTimeout,
+		)
+		defer cancel()
+
+		_, err := c.GetAssetsUnlockedEvents(
+			ctxWithTimeout,
+			sdkmath.NewInt(1),
+			sdkmath.NewInt(2),
+		)
+		if err != nil {
+			logger.Error(
+				"ethereum sidecar bridge-out connection test failed; possible " +
+					"problem with sidecar configuration or connectivity",
+			)
+		} else {
+			logger.Info(
+				"ethereum bridge-out sidecar connection test completed " +
+					"successfully",
+			)
+		}
+	}()
+
+	return c, nil
+}
+
+// GetAssetsUnlockedEvents gets the AssetsUnlocked events from the Mezo chain.
+// The requested range of events is inclusive on the lower side and exclusive
+// on the upper side.
+func (c *BridgeOutGrpcClient) GetAssetsUnlockedEvents(
+	ctx context.Context,
+	sequenceStart sdkmath.Int,
+	sequenceEnd sdkmath.Int,
+) ([]bridgetypes.AssetsUnlockedEvent, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
```

Comment:
I may be missing something but why do we need a mutex here?

Reply:
Removed the mutex in d2f84005e08737970de7d06e8d9b1d5f5e626ab3.
It's not needed, indeed.